src/mapgen/mg_biome.cpp

```diff
@@ -135,13 +123,16 @@ BiomeGenOriginal::BiomeGenOriginal(BiomeManager *biomemgr,
 	// is disabled.
 	memset(biomemap, 0, sizeof(biome_t) * m_csize.X * m_csize.Z);
 
-	// Calculate cache of Y transition points
+	// Calculate cache of Y transition points. For downwards scanning only!
 	std::vector<s16> values;
 	values.reserve(m_bmgr->getNumObjects() * 2);
 	for (size_t i = 0; i < m_bmgr->getNumObjects(); i++) {
 		Biome *b = (Biome *)m_bmgr->getRaw(i);
 		values.push_back(b->max_pos.Y);
-		values.push_back(b->min_pos.Y);
+		// Optimization: biomes often have perfect Y transitions, resulting in
+		// unnecessary Y pairs. i.e. when (biome A min Y) == (biome B max Y + 1)
```

Comment:
i now understand what the code does, but isn't the comment misleading?
it's not about optimization but rather if you have a biome from [min, max] then the points at which a different biome starts are:
* max
* min-1

putting `min` in there is just wrong as the biome at min is equal to min+1

Reply:
Currently, this code could handle both cases: top->down (high Y to low Y) and bottom->up scanning.
We're only using top->bottom so far, thus I'll instead comment this "limitation" of `m_transitions_y` in the header.

Reply:
yeah it was only meant to do top-down from the start